src/nvim/event/socket.c

```diff
@@ -168,3 +169,71 @@ static void close_cb(uv_handle_t *handle)
     watcher->close_cb(watcher, watcher->data);
   }
 }
+
+static void connect_cb(uv_connect_t *req, int status)
+{
+  int *ret_status = req->data;
+  *ret_status = status;
+}
+
+bool socket_connect(Loop *loop, Stream *stream,
+                    bool tcp, const char *address,
+                    int timeout, const char **error)
+{
+  bool success = false;
+  int status;
+  uv_connect_t req;
+  req.data = &status;
+  uv_stream_t *uv_stream;
+  uv_getaddrinfo_t addr_req;
+  addr_req.addrinfo = NULL;
+  const struct addrinfo *addrinfo = NULL;
+  char *addr = NULL;
+  if (tcp) {
+    addr = xstrdup(address);
+    char *host_end = xstrchrnul(addr, ':');
```

Comment:
What about IPv6 addresses, they contain `:` at different positions? Perhaps search for the last `':'` in an address like `::1:8888`?